src/SAML2/Assertion.php

```diff
@@ -523,96 +496,45 @@ private function parseAttributes(DOMElement $xml): void
-            } else {
-                if ($this->nameFormat !== $nameFormat) {
-                    $this->nameFormat = Constants::NAMEFORMAT_UNSPECIFIED;
-                }
             }
-
-            if (!array_key_exists($name, $this->attributes)) {
-                $this->attributes[$name] = [];
-                $this->attributesValueTypes[$name] = [];
-            }
-
-            $this->parseAttributeValue($attribute, $name);
         }
     }
 
-
     /**
-     * @param \DOMNode $attribute
-     * @param string   $attributeName
+     * Parse attribute statements in assertion.
+     *
+     * @param \DOMElement $xml The XML element with the Attribute.
+     * @param boolean $firstAttribute Whether this is the first attribute to parse.
+     * @throws \Exception
      * @return void
      */
-    private function parseAttributeValue(DOMNode $attribute, string $attributeName): void
-    {
-        /** @var \DOMElement[] $values */
-        $values = Utils::xpQuery($attribute, './saml_assertion:AttributeValue');
-
-        if ($attributeName === Constants::EPTI_URN_MACE || $attributeName === Constants::EPTI_URN_OID) {
-            foreach ($values as $index => $eptiAttributeValue) {
-                /** @var \DOMElement[] $eptiNameId */
-                $eptiNameId = Utils::xpQuery($eptiAttributeValue, './saml_assertion:NameID');
-
-                if (count($eptiNameId) === 1) {
-                    $this->attributes[$attributeName][] = new NameID($eptiNameId[0]);
-                } else {
-                    /* Fall back for legacy IdPs sending string value (e.g. SSP < 1.15) */
-                    Utils::getContainer()->getLogger()->warning(
-                        sprintf("Attribute %s (EPTI) value %d is not an XML NameId", $attributeName, $index)
-                    );
-                    $nameId = new NameID();
-                    $nameId->setValue($eptiAttributeValue->textContent);
-                    $this->attributes[$attributeName][] = $nameId;
-                }
-            }
-
-            return;
+    private function parseAttribute(DOMElement $xml, bool $firstAttribute = false): void
+    {
+        $attribute = new Attribute($xml);
+        
+        if ($attribute->getNameFormat() !== null) {
+            $nameFormat = $attribute->getNameFormat();
+        } else {
+            $nameFormat = Constants::NAMEFORMAT_UNSPECIFIED;
         }
 
-        foreach ($values as $value) {
-            $hasNonTextChildElements = false;
-            foreach ($value->childNodes as $childNode) {
-                /** @var \DOMNode $childNode */
-                if ($childNode->nodeType !== XML_TEXT_NODE) {
-                    $hasNonTextChildElements = true;
-                    break;
-                }
-            }
-
-            $type = $value->getAttribute('xsi:type');
-            if ($type === '') {
-                $type = null;
+        if ($firstAttribute) {
+            $this->nameFormat = $nameFormat;
```

Comment:
Why are we still keeping a NameFormat value on the Assertion-class?

Reply:
I didn’t want to rufffle the feathers too much. I’ll leave that up to you guys who “own” the codebase. I just wanted to make the tests pass. 
I’m not a huge fan of completely breaking BC but I must admit all of these changes kinda are anyway so I guess if you want to remove it, now would be the time. It really only makes sense to have when generating an assertion anyway. It just means that there will be no way to apply a single nameformat to all attributes at once but perhaps this functionality can be pushed out to SSP if it’s still needed